components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging-mechanism.adoc

```diff
@@ -0,0 +1,197 @@
+
+Note that this flow should only be used in edge cases such as bugs on the Layer 2 contract preventing message consumption.
+
+[id="l1-l2-message-fees"]
+=== L1 → L2 message fees
+
+An L1 → L2 message induces a transaction on L2, which, unlike regular transactions, is not sent by an account. This calls for a different mechanism for paying the transaction's fee, for otherwise the sequencer has no incentive of including L1 handler transactions inside a block.
+
+To avoid having to interact with both L1 and L2 when sending a message, L1 → L2 messages are payable on L1, by sending ETH with the call to the payable function `sendMessageToL2` on the Starknet Core Contract.
+
+The sequencer takes this fee in exchange for handling the message. The sequencer charges the fee in full upon updating the L1 state with the consumption of this message.
+
+The fee itself is calculated in the xref:../Network_Architecture/fee-mechanism.adoc#overall_fee[same manner] as
+"regular" L2 transactions. You can use the xref:documentation:cli:starkli.adoc#starknet-estimate_fee[CLI] to get an estimate of an L1 → L2 message fee.
+
+[id="structure_and_hashing_l1-l2"]
+=== L1 → L2 structure and hashing
+
+For completeness, we describe the precise structure of both the message as it appears on L1 and the induced transaction as it appears on L2.
+
+.L1 → L2 Message
+[%autowidth.stretch]
+|===
+| FromAddress       | ToAddress      | Selector       | Payload              | Nonce          |
+
+| `EthereumAddress` | `FieldElement` | `FieldElement` | `List+++<FieldElement>+++` | `FieldElement` |
+|===
+
+The hash of the message is computed on L1 as follows:
+
+[source,js]
+----
+keccak256(
+    abi.encodePacked(
+        uint256(FromAddress),
+        ToAddress,
+        Nonce,
+        Selector,
+        Payload.length,
+        Payload
+    )
+);
+----
+
+.L1 handler transaction
+
+[%autowidth.stretch]
+|===
+| Version        | ContractAddress | Selector             | Calldata       | Nonce          |
+
+| `FieldElement` | `FieldElement`  | `FieldElement` | `List+++<FieldElement>+++` | `FieldElement` |
+|===
+
+The hash of the corresponding L1 handler transaction on L2 is computed as follows:
+
+[source,cairo]
+----
+l1_handler_tx_hash = ℎ(
+    "l1_handler",
+    version,
+    contract_address,
+    entry_point_selector,
+    ℎ(calldata),
+    chain_id,
+    nonce
+)
+----
+
+Where:
+
+- stem:[\text{l1_handler}] is a constant prefix, encoded in bytes (ASCII), with big-endian.
+- stem:[\text{chain_id}] is a constant value that specifies the network to which this transaction is sent.
+- stem:[$$h$$] is the xref:../Cryptography/hash-functions.adoc#pedersen_hash[Pedersen] hash
+
+[NOTE]
+====
+In an `l1_handler` transaction, the first element of the calldata is always the Ethereum address of the sender.
```

Comment:
```suggestion
In an `l1_handler` transaction, the first element of the calldata is always the sender's Ethereum address.
```

Reply:
Fixed in e5cf97f8ae042005219bf36f22436d21f869e44f.